### PR TITLE
Docs: aws_lambda_permission clarifications.

### PR DESCRIPTION
**Pull request by SyntaxColoring**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

# Summary

This PR makes two tweaks to the [aws_lambda_permission documentation](https://www.terraform.io/docs/providers/aws/r/lambda_permission.html).

## First (more important)

Rewrite the `source_arn` documentation to clarify what happens when that attribute isn't provided.

I'm assuming Terraform's `aws_lambda_permission.source_arn` is analogous to CloudFormation's `AWS::Lambda::Permission.SourceArn`. [Buried in the CloudFormation docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html), there's a note:

> If you grant permission to a service principal without specifying the source [with `SourceArn`], other accounts could potentially configure resources in their account to invoke your Lambda function.

This surprised me, so it might surprise other people too. Since it's a security implication, we should err towards caution and elaborate on why the attribute is necessary and how to use it properly.

## Second (less important)

Rewrite the summary at the top of the page for grammatical-ness and readability.

# Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md)

```release-note
Docs: Clarified security implications of aws_lambda_permission.source_arn.
```

### CODE REVIEW

File: website/docs/r/lambda_permission.html.markdown [hunk 1]
```diff
@@ -141,16 +140,16 @@ resource "aws_lambda_permission" "lambda_permission" {
...
+ * `source_arn` - (Optional) When the principle is an AWS service, the ARN of the specific resource within that service to grant permission to.
+  Without this, any resource from `principle` will be granted permission – even if that resource is from another account.
+  For S3, this should be the ARN of the S3 Bucket.
+  For CloudWatch Events, this should be the ARN of the CloudWatch Events Rule.
+  For API Gateway, this should be the ARN of the API, as described [here][2].
  * `statement_id` - (Optional) A unique statement identifier. By default generated by Terraform.
  * `statement_id_prefix` - (Optional) A statement identifier prefix. Terraform will generate a unique suffix. Conflicts with `statement_id`.
 
 [1]: https://developer.amazon.com/docs/custom-skills/host-a-custom-skill-as-an-aws-lambda-function.html#use-aws-cli
+[2]: http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html.
```
Comment (anGie44) [inline website/docs/r/lambda_permission.html.markdown:152]: link unfortunately doesn't open correctly with the ending `.` but if removed, works as expected. ah also do you mind changing this to `https`? will be helpful for when AWS officially decides to no longer accept `http` for this page

File: website/docs/r/lambda_permission.html.markdown [hunk 2]
```diff
@@ -141,16 +140,16 @@ resource "aws_lambda_permission" "lambda_permission" {
...
  	e.g. `arn:aws:lambda:aws-region:acct-id:function:function-name:2`
  * `source_account` - (Optional) This parameter is used for S3 and SES. The AWS account ID (without a hyphen) of the source owner.
- * `source_arn` - (Optional) When granting Amazon S3 or CloudWatch Events permission to
- 	invoke your function, you should specify this field with the Amazon Resource Name (ARN)
- 	for the S3 Bucket or CloudWatch Events Rule as its value.  This ensures that only events
- 	generated from the specified bucket or rule can invoke the function.
- 	API Gateway ARNs have a unique structure described
- 	[here](http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html).
+ * `source_arn` - (Optional) When the principle is an AWS service, the ARN of the specific resource within that service to grant permission to.
+  Without this, any resource from `principle` will be granted permission – even if that resource is from another account.
```
[the remark this answers is not in the document]
Comment (SyntaxColoring) [reply website/docs/r/lambda_permission.html.markdown]: Hmm. So you're saying "configured to invoke a function" doesn't necessarily mean "able to invoke a function", in this case?

I admit I haven't actually tested this. I'll try to test it to make sure my understanding is correct.

To confirm, it's that we're uncertain about the behavior, right? Not merely my word choice? For example, it wouldn't be better if I said:

> Without this, any resource from `principal` will ~be granted permission~ **be able to invoke your function.**

[the remark this answers is not in the document]
Comment (anGie44) [reply website/docs/r/lambda_permission.html.markdown]: > "configured to invoke a function" doesn't necessarily mean "able to invoke a function", in this case?

yep, that was my initial interpretation but i'm with you in that I don't have experience to confirm one way or another  😅 i'd say to err on the side of caution, you can use AWS' wording so we don't introduce extra room for interpretation from the provider perspective
```
If you grant permission to a service principal without specifying the source, other accounts could potentially configure resources in their account to invoke your Lambda function.
```

[the remark this answers is not in the document]
Comment (SyntaxColoring) [reply website/docs/r/lambda_permission.html.markdown]: I just played with it a bit, and confirmed that my original understanding was correct.

I set up an S3 bucket on one AWS account, and configured one of its events to invoke a Lambda function on a second AWS account. All I needed was the function's ARN.

With `source_arn` unset, the function executed on each event, even though it came from a different account.

With `source_arn` set to the ARN of a bucket, the function only executed on events from that bucket.

With this in mind, I think my wording is clearer than Amazon's. But I'm happy to keep working on it, if you see more room for improvement.

[the remark this answers is not in the document]
Comment (anGie44) [reply website/docs/r/lambda_permission.html.markdown]: awesome, thanks for investigating @SyntaxColoring! let's go with your wording as is then, makes sense to document w/your findings as I agree the AWS docs leave things a bit unclear

**Pull request merged**